torch/distributed/_sharded_tensor/__init__.py

```diff
@@ -348,3 +360,121 @@ def _recurse_update_dict(module, destination, prefix):
     for submodule_name, submodule in module.named_modules():
         if submodule_name != '':
             _recurse_update_dict(submodule, destination, prefix + submodule_name + '.')
+
+def shard_parameter(
+        module: torch.nn.Module,
+        param_name: str,
+        sharding_spec: ShardingSpec,
+        src_rank=0,
+        process_group=None):
+    """
+    Given a :class:`torch.nn.Module`, a ``param_name`` for a parameter in that
+    module, it shards that parameter according to the provided
+    ``sharding_spec``. ``src_rank`` denotes the source rank which would be
+    used as the ground truth of the data which would be scattered as shards
+    across the rest of the ranks.
+
+    This method replaces ``module.param_name`` with a
+    :class:`torch.distributed._sharded_tensor.ShardedTensor`
+
+    Args:
+        module (:class:`torch.nn.Module`): Module whose parameter needs to be sharded.
+        param_name (str): Name of the parameter of ``module`` that needs to be sharded.
+        sharding_spec (:class:`torch.distributed._sharding_spec.ShardingSpec`): The specification
+            describing how to shard the Tensor.
+        size (int...): a sequence of integers defining the shape of the output
+            tensor. Can be a variable number of arguments or a collection like a list or tuple.
+
+    Keyword args:
+        src_rank (int, optional): The source rank which is used as the ground truth of
+            the data that would be sharded and scattered across the rest of the ranks.
+            Default: 0.
+        process_group (ProcessGroup, optional): The process group to work on. If None,
+            the default process group will be used.
+    """
+    # Perform some validation first.
+    if not isinstance(sharding_spec, ChunkShardingSpec):
```

Comment:
I am thinking that, instead of providing an universal `shard_parameter` API that will shard parameter of a passed in module, and supporting limited set of ShardingSpec, should we add the API like `ShardingSpec.shard(module, param_name, ..)`? This way we might have more flexibility for user to extend it, i.e. ChunkShardingSpec will have the implementation here, other ShardingSpec or use specified ShardingSpec could implement their own sharding strategies.

Reply:
interesting discussion, probably we can introduces something like 

Module.shard 
or Sharder.shard 
or sth else? 

basically try to have a way to let each module owner to customizer the actual sharding if needed. 

Probably add a TODO somewhere to revisit since this shard API of module is very critical ?